frontend/src/core/codemirror/rtc/extension.ts

```diff
@@ -1,22 +1,115 @@
 /* Copyright 2024 Marimo. All rights reserved. */
-import { yCollab } from "y-codemirror.next";
+
+import { Awareness, LoroDoc, LoroText } from "loro-crdt";
 import type { CellId } from "@/core/cells/ids";
 import { isWasm } from "@/core/wasm/utils";
 import type { Extension } from "@codemirror/state";
-import { CellProviderManager } from "./cell-manager";
+import ReconnectingWebSocket from "partysocket/ws";
+import { createWsUrl } from "@/core/websocket/createWsUrl";
+import { getSessionId } from "@/core/kernel/session";
+import { once } from "@/utils/once";
+import { LoroSyncPlugin } from "./loro/sync";
 import { EditorView, ViewPlugin } from "@codemirror/view";
 import {
+  getInitialLanguageAdapter,
   languageAdapterState,
   setLanguageAdapter,
   switchLanguage,
 } from "../language/extension";
 import { LanguageAdapters } from "../language/LanguageAdapters";
 import type { LanguageAdapterType } from "../language/types";
+import { atom } from "jotai";
+import { waitForConnectionOpen } from "@/core/network/connection";
+import { store } from "@/core/state/jotai";
+import { getFeatureFlag } from "@/core/config/feature-flag";
+import { initialMode } from "@/core/mode";
 import { Logger } from "@/utils/Logger";
 
+export const connectedDocAtom = atom<LoroDoc | undefined>(undefined);
+
+// Create a Loro document
+const doc = new LoroDoc();
+const awareness: Awareness = new Awareness(doc.peerIdStr);
+// const undoManager = new UndoManager(doc, {});
+
+// Create a websocket connection to the server
+const getWs = once(() => {
+  Logger.debug("[rtc] creating websocket");
+
+  const url = createWsUrl(getSessionId()).replace("/ws", "/ws_sync");
+
+  // Create the websocket, but don't connect it yet
+  const ws = new ReconnectingWebSocket(url, undefined, {
+    // We don't want Infinity retries
+    maxRetries: 10,
+    debug: false,
+    startClosed: true,
+    connectionTimeout: 10_000,
+  });
+
+  // Start the connection
+  Promise.resolve().then(async () => {
+    // First wait until the connection is open
+    await waitForConnectionOpen();
+
+    // Now open the websocket
+    ws.reconnect();
+  });
+
+  // Receive updates from the server
+  ws.addEventListener("message", async (event) => {
+    const blob: Blob = event.data;
+    const bytes = await blob.arrayBuffer();
+    doc.import(new Uint8Array(bytes));
+    Logger.debug("[rtc] imported doc change. new doc:", doc.toJSON());
+
+    // Set the active doc
+    if (store.get(connectedDocAtom) !== doc) {
+      store.set(connectedDocAtom, doc);
+    }
+  });
+
+  // Handle open event
+  ws.addEventListener("open", () => {
+    Logger.debug("[rtc] websocket open");
+  });
+
+  // Handle close event
+  ws.addEventListener("close", () => {
+    Logger.debug("[rtc] websocket close");
+
+    // Remove the active doc
+    store.set(connectedDocAtom, undefined);
+  });
+
+  return ws;
+});
+
+// Kick off the connection for edit mode
+if (getFeatureFlag("rtc_v2") && initialMode === "edit") {
```

Comment:
Maybe worth having a different feature flag? really v3

Reply:
i think its ok. the reason for switching it was that i wanted to force turn everyone else off, since it created issues and was in the Labs UI. this is an improvement and we have't advertised rtc_v2